File: discord.py/games/hangman.py
from discord import Embed, Color, Message
from discord.ext.commands import Context

from core.cls import Bot
from random import choice
from core.tools import normalize_string


class Hangman:
    def __init__(self, bot, ctx):
        self.bot: Bot = bot
        self.ctx: Context = ctx
        self.message: Message = None
        self.embed: Embed = None

        self.word = choice([ligne.strip() for ligne in open('wordlist.txt', encoding='UTF-8')])
        self.normalized = normalize_string(self.word)
        self.guess = ['-']*len(self.word)
        self.lives, self.errors = 5, []

    async def start(self) -> None:
        self.embed = (Embed(title='Partie de pendu', color=Color.random())
                      .add_field(name='Mot', value=f"```{''.join(self.guess)}```", inline=False)
                      .add_field(name='Erreurs', value='```\u200b```', inline=False)
                      .set_footer(text=f'Vies : {self.lives}'))

        self.message = await self.ctx.send(embed=self.embed)
        await self.play()

    async def get_letter(self) -> str:
        message = await self.bot.wait_for('message', check=lambda m: m.author == self.ctx.author and len(m.content) in [1, len(self.word)])
        await message.delete()
        return normalize_string(message.content.lower())

    async def play(self) -> None:
        letter = await self.get_letter()

        if letter in self.errors or letter in self.guess:
            await self.ctx.send('Tu as déjà envoyé cette lettre')
            return await self.play()

        if len(letter) in [1, len(self.word)] and letter not in self.normalized and letter != self.normalized:
            self.lives -= 1
            self.errors.append(letter)
            self.embed.set_field_at(1, name='Erreurs', value=f"```{', '.join(self.errors)}```", inline=False)
            self.embed.set_footer(text=f'Vies : {self.lives}')
        elif len(letter) == 1:
            self.guess = [self.word[i] if letter == char else self.guess[i] for i, char in enumerate(self.normalized)]
            self.embed.set_field_at(0, name='Mot', value=f"```{''.join(self.guess)}```", inline=False)

        if self.word == ''.join(self.guess) or self.normalized == letter:
            self.embed.set_field_at(0, name='Mot', value=f"```{self.word}```", inline=False)
            await self.message.edit(embed=self.embed)
            return await self.ctx.send(f'Bravo, tu as gagné ! :)')

        if not self.lives:
            return await self.ctx.send(f'Perdu ! Le mot était `{self.word}`')

        await self.message.edit(embed=self.embed)
        await self.play()

File: JDA PolyBot/src/main/kotlin/commands/games/Hangman.kt
package commands.games

import net.dv8tion.jda.api.EmbedBuilder
import net.dv8tion.jda.api.entities.MessageEmbed
import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent
import net.dv8tion.jda.api.events.message.MessageReceivedEvent
import net.dv8tion.jda.api.hooks.ListenerAdapter
import net.dv8tion.jda.api.interactions.InteractionHook
import net.dv8tion.jda.api.interactions.commands.build.Commands
import java.io.File
import java.text.Normalizer
import java.util.*
import kotlin.collections.ArrayList

object Hangman {
    val commandData = Commands.slash("pendu", "Jouer au pendu")

    fun execute(event: SlashCommandInteractionEvent) = Listener(event)

    class Listener(private val event: SlashCommandInteractionEvent): ListenerAdapter() {
        private lateinit var hook: InteractionHook
        private val word = File("wordlist.txt").readLines().shuffled()[0]

        private var lives = 5
        private val normalized = normalize(word)
        private val errors = ArrayList<String>()
        private var discovered = Array(word.length) {false}

        init {
            event.interaction.replyEmbeds(buildEmbed()).queue {
                hook = it
                event.jda.addEventListener(this)
            }
        }

        override fun onMessageReceived(event: MessageReceivedEvent) {
            if (hook.interaction.user != event.author) return
            if (hook.interaction.messageChannel.idLong != event.channel.idLong) return

            val guess = normalize(event.message.contentRaw).lowercase(Locale.getDefault())
            event.message.delete().queue()

            if (guess in errors) return

            if (lives == 1) {
                hook.jda.removeEventListener(this)
                updateEmbed(EmbedBuilder().setColor(Colors.RED).setDescription("Perdu ! Le mot était  `$word`").build())
                return
            }

            if (guess != normalized && guess !in normalized) {
                lives--
                errors.add(guess)
                updateEmbed()
                return
            }

            word.forEachIndexed pass@{i, c ->
                if (discovered[i]) return@pass
                discovered[i] = c == guess[0]
            }

            if (guess.length > 1 || discovered.all{b -> b}) {
                hook.jda.removeEventListener(this)
                updateEmbed(EmbedBuilder().setColor(Colors.GOLD).setDescription("Gagné ! Tu as deviné `$word`").build())
                return
            }

            updateEmbed()
        }

        private fun updateEmbed(embed: MessageEmbed? = null) {
            hook.editOriginalEmbeds(embed ?: buildEmbed()).queue()
        }

        private fun buildEmbed(): MessageEmbed {
            val errorsStr = if (errors.isEmpty()) "\u200b" else errors.joinToString(", ")
            val wordStr = word.mapIndexed{i, c -> if(discovered[i]) c else "-"}.joinToString("")

            return EmbedBuilder().setTitle("Partie de pendu")
                .addField("Mot : ", "```$wordStr```", false)
                .addField("Erreurs", "```$errorsStr```", false)
                .setFooter("Vies : $lives")
                .setColor(Colors.BLUE)
                .build()
        }

        private fun normalize(word: String): String {
            return Normalizer.normalize(word, Normalizer.Form.NFKD)
        }
    }
}